templates/fine-tune-function-calling/fc_utils/eval_utils.py

```diff
@@ -0,0 +1,163 @@
+                    and current_conv[-2]["tool_calls"]
+                ):
+                    current_conv[-1]["tool_call_id"] = current_conv[-2]["tool_calls"][0].id
+                else:
+                    return current_conv, False
+            if isinstance(parser, OpenAIResponseParser):
+                parsed_response = parser.get_parsed_response(current_conv, tools)
+            else:
+                parsed_response = parser.get_parsed_response(current_conv)
+            if parsed_response["content"] == ERROR_OUTPUT:
+                return None, None  # return None if there's an error
+
+            is_correct, mistake_type = is_match(parsed_response, expected_responses[assist_id])
+            match = match and (is_correct)
+            original_assistant_response = parsed_response["original_response"]
+            current_conv.append(dict(original_assistant_response))
+            assist_id += 1 # next expected response
+            if not match:  # return right away if model output is incorrect
+                return current_conv, match, mistake_type
+        conv_id += 1 # next message
+    return current_conv, match, mistake_type
+
+
+def evaluate_gpt(ds: List[dict], openai_parser: OpenAIResponseParser):
+    """
+    Evaluates the GPT4 model on the test dataset.
+    """
+    num_correct = 0
+    total_count = 0
+    results = []
+    pbar = tqdm(total=len(ds), desc="Evaluating GPT4..")
+    for example in ds:
+        openai_messages = example["openai_messages"]
+        openai_tools = [
+            {"type": "function", "function": fn} for fn in example["tools"]
+        ]
+        openai_conv, openai_is_match, mistake_type = parse_and_eval(
+            openai_parser, openai_messages, example["expected_responses"], tools=openai_tools
+        )
+        pbar.update(1)
+        if openai_conv is None:  # skip if api errors out
+            openai_is_match = None
+        else:
+            total_count += 1
+        result_dict = {"correct": openai_is_match, "mistake_type": mistake_type, "example": example, "conv": openai_conv}
+        results.append(result_dict)
+
+        if openai_is_match:
+            num_correct += 1
+    print("GPT4 accuracy: ", num_correct / total_count)
+    return results
+
+def evaluate_finetuned(ds: List[dict], anyscale_parser: AnyscaleResponseParser):
+    """
+    Evaluates our fine-tuned model on the test dataset.
+    """
+    anyscale_accuracy = 0
+    total_count = 0
+    results = []
+    pbar = tqdm(total=len(ds), desc="Evaluating Finetuned Model..")
+    for example in ds:
+        anyscale_messages = example["anyscale_messages"]
+        anyscale_conv, anyscale_is_match, mistake_type = parse_and_eval(anyscale_parser, anyscale_messages, example["expected_responses"])
+        pbar.update(1)
+        if anyscale_conv is None: # skip if api errors out
+            anyscale_is_match = None
+        else:
+            total_count += 1
```

Comment:
unify the return types. Don't mix up None with empty string, etc. Or at least define the interface more consistently.